### ISSUE TITLE
Calculating median cast integers as floats

### ISSUE THREAD

**Issue by mortengan** [pola-rs/polars#17125]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python

import polars as pl

df = pl.DataFrame(data=[int(1e18), int(1e18) + 1, int(1e18) + 2])

print(df)

print(df.select(pl.col('column_0').median()))

print(df.select(pl.col('column_0').median().cast(int)))
```

### Log output

```shell
shape: (3, 1)
┌─────────────────────┐
│ column_0            │
│ ---                 │
│ i64                 │
╞═════════════════════╡
│ 1000000000000000000 │
│ 1000000000000000001 │
│ 1000000000000000002 │
└─────────────────────┘
shape: (1, 1)
┌───────────┐
│ column_0  │
│ ---       │
│ f64       │
╞═══════════╡
│ 1.0000e18 │
└───────────┘
shape: (1, 1)
┌─────────────────────┐
│ column_0            │
│ ---                 │
│ i64                 │
╞═════════════════════╡
│ 1000000000000000000 │
└─────────────────────┘
```

### Issue description

When using integers, calculating median returns a float, which in case of large numbers results in a wrong calculation, as float64 does not have the same precission. 

### Expected behavior

At least, given an odd number of elements, it should return an integer, which in this case is should be `1000000000000000001`.

### Installed versions

<details>

```
--------Version info---------
Polars:               0.20.31
Index type:           UInt32
Platform:             Linux-6.8.0-35-generic-x86_64-with-glibc2.39
Python:               3.11.9 (main, Apr 28 2024, 18:52:29) [GCC 13.2.0]

----Optional dependencies----
adbc_driver_manager:  <not installed>
cloudpickle:          <not installed>
connectorx:           <not installed>
deltalake:            <not installed>
fastexcel:            <not installed>
fsspec:               <not installed>
gevent:               <not installed>
hvplot:               <not installed>
matplotlib:           <not installed>
nest_asyncio:         <not installed>
numpy:                1.26.4
openpyxl:             <not installed>
pandas:               <not installed>
pyarrow:              <not installed>
pydantic:             <not installed>
pyiceberg:            <not installed>
pyxlsb:               <not installed>
sqlalchemy:           <not installed>
torch:                <not installed>
xlsx2csv:             <not installed>
xlsxwriter:           <not installed>
None
```

</details>

**Comment by JulianCologne** [comment]
This is the intended behavior for median calculations.

You already answered your own question. 
For odd-size data the median is calculated by the center elements (c1+c2)/2 and can result in a floating point value.

Polars makes sure (and probably almost all other languages) that the resulting type is consistent regardless of the input.

Resulting an integer sometimes and sometimes a float is problematic.

I get your point with very large floats imprecision but what is your ideal solution?
Maybe some kind of parameter to take the lower/upper value for odd-sized data?

**Comment by mortengan** [comment]
I think having a warning would be necessary, as the output could break assumptions that had been verified earlier. Example:

```
import polars as pl

df = pl.DataFrame(data=[int(1e18) + 1, int(1e18) + 2, int(1e18) + 3])
df = df.with_columns(
    val_test=pl.col('column_0') > int(1e18))
assert True in df.select(pl.col("val_test").all()).row(0)

print(df.select(pl.col('column_0').median()))
```
The assert would not trigger, yet I can receive a median value that does not exist (int(1e18) straight) and was specifically tested to not exist. 

Enforcing no conversion of type when a parameter choosing the larger/smaller number, could be a fix. 

Returning an integer would at any rate also help not getting into trouble later on - if you want to subtract the median from the column from which it was calculated (to get a "difference from median" column), you could end up in all kinds of trouble, as you now are limited to a much lower resolution. You may know that your original column was unique, and subtracting the same number (the median fx) from all values in it would also result in a unique column. Which is wrong given integers of a certain size.

That is not to say that there is no value in being consistent with other libraries - which is why I think that only a warning would be necessary. My personal opinion is that it should never change type for the reasons stated above - but I am not a core programmer, so take it with a pinch of salt :)

**Comment by stinodego** [comment]
This is not a bug as explained by @JulianCologne. We will not throw a warning here, because this is expected behavior. When working with floats, you know that you must account for precision loss.

However, there might be some merit in a dtype-preserving version of some aggregations, so I'll rebrand this issue as a feature request.

As a workaround/approximation, you can sort the column and take the middle value.

**Comment by mortengan** [comment]
@stinodego good to know that it is indeed intended behaviour. If I am the only one who would like a different behavior, nothing should be done about it - obviously.

I don't have any need to keep this issue open unless someone wants further discussions on this... Maybe it is supposed to stay open, given that it is now a feature request. 

Either way thanks for the responses.

**Comment by avimallu** [comment]
> However, there might be some merit in a dtype-preserving version of some aggregations

It might also make sense to have a mention in the docs for [median](https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.median.html) that it can change the datatype like the mention in [mode](https://docs.pola.rs/api/python/stable/reference/series/api/polars.Series.mode.html). Right now, the `Expr -> Expr` type hint doesn't really help.

**Comment by stinodego** [comment]
> It might also make sense to have a mention in the docs for [median](https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.median.html) that it can change the datatype like the mention in [mode](https://docs.pola.rs/api/python/stable/reference/series/api/polars.Series.mode.html). Right now, the `Expr -> Expr` type hint doesn't really help. I can add a PR if this is determined to be helpful.

Yes, feel free to add a `Returns` section specifying the expected return type. Most of the namespace methods already have this, e.g. https://docs.pola.rs/api/python/stable/reference/expressions/api/polars.Expr.str.concat.html
